src/Cornelis/Diff.hs

```diff
@@ -0,0 +1,87 @@
+-- | Maintain a Diff between the text that Agda loaded most recently
+-- and the current Vim buffer. There is one such diff for every buffer,
+-- indexed by 'BufferNum'.
+--
+-- This exports three functions:
+-- - 'resetDiff' empties the diff when reloading Agda
+-- - 'recordUpdate' adds a buffer update to the diff.
+-- - 'translateInterval' applies the Diff to an interval coming from Agda,
+--   turning it into an interval for the current buffer.
+module Cornelis.Diff
+  ( resetDiff
+  , translateInterval
+  , recordUpdate
+  , Replace(..)
+  , Colline(..)
+  , Vallee(..)
+  ) where
+
+import Data.IORef (atomicModifyIORef')
+import qualified Data.Map as Map
+import Data.Tuple (swap)
+import DiffLoc (Replace(..), Colline(..), Vallee(..))
+import qualified DiffLoc as D
+import Neovim
+import Cornelis.Offsets
+import Cornelis.Types
+
+-- | A general function for modifying Diffs, shared between the three functions below.
+modifyDiff :: BufferNum -> (Diff0 -> (Diff0, a)) -> Neovim CornelisEnv a
+modifyDiff buf f = do
+  csRef <- asks ce_state
+  liftIO $ atomicModifyIORef' csRef $ \cs ->
+    let (a, ds') = Map.alterF (fmap Just . swap . alter) buf (cs_diff cs)
+    in (cs { cs_diff = ds' }, a)
+  where
+    alter Nothing = f D.emptyDiff
+    alter (Just d) = f d
+
+-- These three functions are essentially monadic wrappers, using 'modifyDiff',
+-- around the corresponding diff-loc functions:
+--
+-- @
+-- 'D.emptyDiff' :: Diff0
+-- 'D.addDiff' :: D.Replace DPos -> Diff0 -> Diff0
+-- 'D.mapDiff' :: Diff0 -> D.Interval DPos -> Maybe (D.Interval DPos)
+-- @
+--
+-- This relies on instances of 'D.Amor' and 'D.Origin' implemented in 'Cornelis.Offsets'.
+
+-- | Reset the diff to an empty diff.
+resetDiff :: BufferNum -> Neovim CornelisEnv ()
+resetDiff buf = modifyDiff buf $ \_ -> (D.emptyDiff, ())
+
+-- | Add a buffer update (insertion or deletion) to the diff.
+-- The buffer update event coming from Vim is structured exactly how the diff-loc
+-- library expects it.
+recordUpdate :: BufferNum -> D.Replace DPos -> Neovim CornelisEnv ()
+recordUpdate buf r = modifyDiff buf $ \d -> (D.addDiff r d, ())
+
+-- | Given an interval coming from Agda (a pair of start and end positions),
+-- find the corresponding interval in the current buffer.
+-- If an edit touches the interval, return Nothing.
+translateInterval :: BufferNum -> Interval VimPos -> Neovim CornelisEnv (Maybe (Interval VimPos))
```

Comment:
Should we have a type-level indexing mode for this? I have fears I'm going to forget to run this function and be surprised when my intervals are in the wrong place.

Reply:
I also feel anxious about that already :) I wonder if we can keep it simple and say that `AgdaPos` is always an old position, and a `VimPos` is always a new position, so this `translateInterval` logic would be used in the only function that lets us translate `AgdaPos` to `VimPos`.

Reply:
I'm not sure that works; when we communicate to Agda we need to send the current versions of Agda intervals.

Reply:
I can also write the inverse translation function, but it seems that the only intervals we ever need to send to Agda are goals, whose location is provided by Agda already.